File: src/2025-02-24-deepclaude-rust/go/exercises/practice/pig-latin/pig_latin.go
package piglatin

import "strings"

func isVowel(c rune) bool {
	return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u'
}

func allConsonants(s string) bool {
	for _, c := range s {
		if isVowel(c) {
			return false
		}
	}
	return true
}

func translateWord(word string) string {
	// Rule 1: starts with vowel, xr, yt
	if len(word) >= 2 {
		if word[:2] == "xr" || word[:2] == "yt" {
			return word + "ay"
		}
	}
	if len(word) >= 1 && isVowel(rune(word[0])) {
		return word + "ay"
	}

	// Rule 3: check for "qu" after consonants
	for i := 0; i < len(word)-1; i++ {
		if i+2 <= len(word) && word[i:i+2] == "qu" {
			prefix := word[:i]
			if allConsonants(prefix) {
				return word[i+2:] + prefix + "qu" + "ay"
			}
		}
	}

	// Rule 4: consonants followed by y
	for i := 0; i < len(word); i++ {
		if word[i] == 'y' {
			prefix := word[:i]
			if len(prefix) >= 1 && allConsonants(prefix) {
				return word[i:] + prefix + "ay"
			}
		}
	}

	// Rule 2: move consonants to end
	for i, c := range word {
		if isVowel(c) {
			return word[i:] + word[:i] + "ay"
		}
	}
	return word + "ay"
}

func Sentence(sentence string) string {
	words := strings.Split(sentence, " ")
	translated := make([]string, len(words))
	for i, word := range words {
		translated[i] = translateWord(word)
	}
	return strings.Join(translated, " ")
}

File: src/python-2025-02-27-deepclaude37-rust/python/exercises/practice/pig-latin/pig_latin.py
def translate(text):
    def get_split_idx(word):
        vowels = {'a', 'e', 'i', 'o', 'u'}
        
        # Rule 1: If word begins with vowel, 'xr', or 'yt'
        if word[0] in vowels or word.startswith('xr') or word.startswith('yt'):
            return 0
            
        # Rule 3: If word has consonants (possibly zero) followed by 'qu'
        for i in range(len(word) - 1):
            if word[i] == 'q' and word[i+1] == 'u':
                return i + 2  # Split after 'qu'
                
        # Find first occurrence of a vowel or 'y'
        first_vowel_or_y = None
        for i, c in enumerate(word):
            if c in vowels or c == 'y':
                first_vowel_or_y = i
                break
        
        # Rule 4: Check if first is 'y' and preceded by consonants
        if first_vowel_or_y is not None and word[first_vowel_or_y] == 'y' and first_vowel_or_y > 0:
            return first_vowel_or_y
        
        # Rule 2: Split at first vowel (if found)
        if first_vowel_or_y is not None:
            return first_vowel_or_y
        
        # Edge case: if no vowels found
        return len(word)
        
    def translate_word(word):
        split_idx = get_split_idx(word)
        return word[split_idx:] + word[:split_idx] + 'ay'
        
    # Process each word in the text
    return ' '.join(translate_word(word) for word in text.split())
